parcels/interaction/interactionkernelsoa.py

```diff
@@ -0,0 +1,179 @@
+
+        # At this time, JIT mode is not supported for InteractionKernels,
+        # so there is no need for any further "processing" of pyfunc's.
+
+    def execute_jit(self, pset, endtime, dt):
+        raise NotImplementedError
+
+    def __del__(self):
+        # Clean-up the in-memory dynamic linked libraries.
+        # This is not really necessary, as these programs are not that large, but with the new random
+        # naming scheme which is required on Windows OS'es to deal with updates to a Parcels' kernel.)
+        super().__del__()
+
+    def __add__(self, kernel):
+        if not isinstance(kernel, InteractionKernelSOA):
+            kernel = InteractionKernelSOA(self.fieldset, self.ptype, pyfunc=kernel)
+        return self.merge(kernel, InteractionKernelSOA)
+
+    def __radd__(self, kernel):
+        if not isinstance(kernel, InteractionKernelSOA):
+            kernel = InteractionKernelSOA(self.fieldset, self.ptype, pyfunc=kernel)
+        return kernel.merge(self, InteractionKernelSOA)
+
+    def execute_python(self, pset, endtime, dt):
+        """Performs the core update loop via Python"""
+        if self.fieldset is not None:
+            for f in self.fieldset.get_fields():
+                if type(f) in [VectorField, NestedField, SummedField]:
+                    continue
+                f.data = np.array(f.data)
+
+        for pyfunc in self._pyfunc:
+            pset.compute_neighbor_tree(endtime, dt)
+            active_idx = pset._active_particle_idx
+
+            mutator = defaultdict(lambda: [])
+            for particle_idx in active_idx:
+                p = pset[particle_idx]
+                # Don't use particles that are not started.
+                if (endtime-p.time)/dt <= -1e-7:
+                    continue
+
+                neighbors = pset.neighbors_by_index(particle_idx)
+                try:
+                    res = pyfunc(p, pset.fieldset, p.time, neighbors, mutator)
+                except FieldOutOfBoundError as fse_xy:
+                    res = ErrorCode.ErrorOutOfBounds
+                    p.exception = fse_xy
+                except FieldOutOfBoundSurfaceError as fse_z:
+                    res = ErrorCode.ErrorThroughSurface
+                    p.exception = fse_z
+                except TimeExtrapolationError as fse_t:
+                    res = ErrorCode.ErrorTimeExtrapolation
+                    p.exception = fse_t
+
+                except Exception as e:
+                    res = ErrorCode.Error
+                    p.exception = e
+
+                # (!) InteractionKernels do not advance time or change state
+                if res != StateCode.Success:
+                    logger.warning_once("Some InteractionKernel was not completed succesfully.")
+
+            for particle_idx in active_idx:
+                p = pset[particle_idx]
+                try:
+                    for mutator_func, args in mutator[p.id]:
```

Comment:
now that looks extension-friendly, using the ID instead of the index of a particle! nice.